### PR TITLE
feat(query): query result cache

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Implement the basic query result cache.

More details about the design can be found in [the RFC](https://databend.rs/doc/contributing/rfcs/query-result-cache).

Pipeline:

```
             ┌─────────┐ 1  ┌─────────┐ 1
             │         ├───►│         ├───►Dummy───►Downstream
Upstream────►│Duplicate│ 2  │         │ 3
             │         ├───►│         ├───►Dummy───►Downstream
             └─────────┘    │         │
                            │ Shuffle │
             ┌─────────┐ 3  │         │ 2  ┌─────────┐
             │         ├───►│         ├───►│  Write  │
Upstream────►│Duplicate│ 4  │         │ 4  │ Result  │
             │         ├───►│         ├───►│  Cache  │
             └─────────┘    └─────────┘    └─────────┘

```

### CODE REVIEW

File: src/query/storages/result_cache/src/write/transform.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    sql: String,
+    partitions_sha: String,
+
+    meta_mgr: ResultCacheMetaManager,
+    cache_writer: ResultCacheWriter,
+}
+
+// The logic is similar to `Transformers`, but use async to handle `on_finish`.
+#[async_trait::async_trait]
+impl Processor for TransformWriteResultCache {
```
[the remark this answers is not in the document]
Comment (zhang2014) [reply src/query/storages/result_cache/src/write/transform.rs:48]:
```
                        +----->Downstream
                        |
                        +----->CacheAsyncSink
Upstream+               |
        |               +----->Downstream
Upstream+------>Broadcas|
        |               +----->CacheAsyncSink
Upstream+               |
                        +----->Downstream
                        |
                        +----->CacheAsyncSink

```

**Comment by RinChanNOWWW** [comment]
**NOTICE**: This PR may be rebased frequently. (on https://github.com/datafuselabs/databend/pull/10098)

**Comment by RinChanNOWWW** [comment]
Almost ready for review.

File: src/query/service/src/sessions/query_ctx.rs [hunk 2]
```diff
@@ -277,6 +279,16 @@ impl TableContext for QueryContext {
         Ok(())
     }
 
+    fn set_paritions_sha(&self, s: String) {
+        let mut sha = self.partitions_sha.write();
```
Comment (sundy-li) [inline src/query/service/src/sessions/query_ctx.rs:283]: What if there are multi tables? The `sha` could be override.

Comment (sundy-li) [reply src/query/service/src/sessions/query_ctx.rs:283]: Need to combine sha by order.

Comment (RinChanNOWWW) [reply src/query/service/src/sessions/query_ctx.rs:283]: Multi tables are not considered now. Will be considered in later PR.

**Comment by RinChanNOWWW** [comment]
Rely on #10127.

File: src/query/service/src/sessions/query_ctx.rs (hunk 2 above)
Comment (sundy-li) [reply src/query/service/src/sessions/query_ctx.rs:283]: If it's not considered, we should have codes to forbid hitting the cache when querying on multi tables.

File: src/query/service/src/interpreters/interpreter_select_v2.rs [hunk 3]
```diff
@@ -65,6 +75,64 @@ impl SelectInterpreterV2 {
...
+        //              │         ├───►│         ├───►│  Cache  │
+        //              └─────────┘    └─────────┘    └─────────┘
+
+        // 1. Duplicate the pipes.
+        pipeline.duplicate(false)?;
+        // 2. Reorder the pipes.
+        let output_len = pipeline.output_len();
+        debug_assert!(output_len % 2 == 0);
+        let mut rule = vec![0; output_len];
+        for (i, r) in rule.iter_mut().enumerate().take(output_len) {
```
Comment (zhang2014) [inline src/query/service/src/interpreters/interpreter_select_v2.rs:99]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3b014b8b2a31d1f2ea26b17d88a1dd62

File: src/query/storages/result_cache/src/common.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+#[inline(always)]
+pub(crate) fn gen_result_cache_dir(key: &str) -> String {
+    format!("{RESULT_CACHE_PREFIX}/{key}")
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+pub(crate) struct ResultCacheValue {
+    /// The query SQL.
+    pub sql: String,
```
Comment (zhang2014) [inline src/query/storages/result_cache/src/common.rs:43]: It's original sql? maybe we should save original sql if not?

Comment (RinChanNOWWW) [reply src/query/storages/result_cache/src/common.rs:43]: It's the origin SQL. It's used for `system.query_result`.

**Review by sundy-li — APPROVED** [review]

**Review by ariesdevil — APPROVED** [review]

File: src/query/storages/result_cache/src/read/reader.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                    // The cache is invalid (due to data update or other reasons).
+                    Ok(None)
+                }
+            }
+            None => Ok(None),
+        }
+    }
+
+    async fn read_result_from_cache(&self, location: &str) -> Result<Vec<DataBlock>> {
+        let object = self.operator.object(location);
```
Comment (BohuTANG) [inline src/query/storages/result_cache/src/read/reader.rs:80]: If the location file is removed from the s3, the query will return an error or?

File: src/query/service/src/interpreters/interpreter_select_v2.rs [hunk 6]
```diff
@@ -81,7 +149,31 @@ impl Interpreter for SelectInterpreterV2 {
...
+            // 1. Try to get result from cache.
+            let kv_store = UserApiProvider::instance().get_meta_store_client();
+            let cache_reader = ResultCacheReader::create(
+                self.ctx.clone(),
+                kv_store.clone(),
+                self.ctx
+                    .get_settings()
+                    .get_tolerate_inconsistent_result_cache()?,
+            );
+            if let Some(blocks) = cache_reader.try_read_cached_result().await? {
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_select_v2.rs:168]: If `try_read_cached_result` returns an error, we can't raise it to the user. Instead, we should log the error and continue the query to make it work.

File: src/query/storages/result_cache/src/read/reader.rs (hunk 5 above)
Comment (RinChanNOWWW) [reply src/query/storages/result_cache/src/read/reader.rs:80]: It will return an error now.

**Pull request merged**